### DISCUSSION TITLE
Should phased restart be removed?

### DISCUSSION THREAD

**Discussion by dentarg** [puma/puma#3034] [Polls]
Is having [phased restart](https://github.com/puma/puma/blob/v6.0.0/docs/restart.md#phased-restart) inside Puma is worth it? 

It is a lot of complexity to maintain and it is becoming an even sharper knife over time (see: [recent problem with Rails and `date` gem](https://github.com/puma/puma/issues/3033#issuecomment-1348791555), https://github.com/puma/puma/issues/2843, https://github.com/puma/puma/issues/2893).

Is it up to Puma to handle deployment strategies?

Just throwing this out there.

**Comment by nateberkopec** [comment]
So, I also don't like maintaining it, but there are several features we maintain like phased-restart which are intended to support a specific use-case of someone who wants to start a Puma server and just wants Puma to provide everything all by itself, without need for additional programs or tools. 

For example:

1. SSL (you could use a reverse proxy like Nginx to terminate your SSL)
2. Request buffering (same as above. Unicorn works like this.)
3. and phased restarts (zero-downtime deployment without something like, say, k8s hot-swapping new and old pods out)

So, to me, the design goal of "it should be safe, easy and effective to run Puma without any other tools" requires supporting these features. Although I would love to get rid of phased restarts, it would really hamper this design goal.

**Reply by dentarg** [reply]
Is it safe though? If we can't maintain it.

**Reply by nateberkopec** [reply]
I think we're doing ok with it. It's also one of the hardest things we have to test, so that's the source of these problems. Maybe we can come up with some more creative solutions regarding testing it.

**Comment by dentarg** [comment]
Connecting this poll with this recent discussion about phased restart: https://github.com/puma/puma/discussions/3236

**Comment by dentarg** [comment]
https://github.com/puma/puma/pull/3477 was an attempt to make phased restarts compatible with [`preload_app!`](https://github.com/puma/puma/blob/450784abb63b841f466e62ad47f440191fcdba54/lib/puma/dsl.rb#L933-L945)
